API/Services/Implementations/SearchService.cs

```diff
@@ -0,0 +1,101 @@
+using API.DTOs;
+using API.Repositories.Interfaces;
+using API.Services.Interfaces;
+
+namespace API.Services.Implementations;
+
+public class SearchService(ITournamentsRepository tournamentsRepository, IMatchesRepository matchesRepository, IPlayerRepository playerRepository) : ISearchService
+{
+    private readonly ITournamentsRepository _tournamentsRepository = tournamentsRepository;
+    private readonly IMatchesRepository _matchesRepository = matchesRepository;
+    private readonly IPlayerRepository _playerRepository = playerRepository;
+
+    public async Task<List<SearchResponseDTO>?> SearchByNameAsync(string? tournamentName, string? matchName, string? username)
+    {
+        IEnumerable<SearchResponseDTO>? returnList = await SearchTournamentsByNameAsync(tournamentName);
+
+        if (returnList is not null)
+        {
+            return returnList.ToList();
+        }
+
+        returnList = await SearchMatchesByNameAsync(matchName);
+
+        if (returnList is not null)
+        {
+            return returnList.ToList();
+        }
+
+        returnList = await SearchPlayersByNameAsync(username);
+
+        if (returnList is not null)
+        {
+            return returnList.ToList();
+        }
+
+        return null;
+    }
```

Comment:
Our list needs to contain all search results, not just the first type of result. Instead of returning early, we should combine all results. So, instead of returning the `returnList` at each if statement, instead extend that list with the results of the queries. So by the end, `returnList` contains results from all searches.

This function can be made non-nullable, just return the empty list of results if nothing's found.

Reply:
Do we want to allow for multiple types of results within a single return? Because each search parameter is for a specific table, I assumed we would only be searching for a single one per request. If that's not the case I think we should probably do either the first option below or both of them:

- Add a new field to `SearchResponseDTO` like `Type` or some similar name
- Potentially merge all 3 name fields into a single field, and search all 3 tables for that name

Reply:
I think how we make the change you suggested and I listed above would depend on where this endpoint is going to be implemented, so I'm curious about exactly where and how we were planning on utilizing it in the website

Reply:
Yeah do both, this is intended to function like a site-wide search. The user clicks the search bar, looking for 'anything' basically. So if they type "Stage", the player stage could appear, a beatmap with [Extra Stage] could appear, etc.

I think for an item-specific search (such as tournament search by itself) we should create separate endpoints for that. Ideally the codebase is structured in a way where it's not complicated to do both combined and distinct type searches.

Reply:
We commented at the exact same time but you answered my question in your comment :ehehe:. That makes sense, I'll implement both of those things into the logic